### ISSUE TITLE
[RLlib] A3C with LSTM not performing in production as during training 

### ISSUE THREAD

**Issue by fpartika** [ray-project/ray#34817]
### What happened + What you expected to happen

I've successfully trained a A3C agent using RLlib with incredible speed and reaching good rewards, unfortunately testing the saved agent with same training environment gave me only 25% of the average training reward, even setting exploration off it is too far from the training result.
Thanks in advance for your support!

### Versions / Dependencies

Tensorflow 2.6.2
RLlib 2.3.0

### Reproduction script

register_env("custom_env", lambda config: Env(config))

env=Env('')
observation_space=env.observation_space
action_space=env.action_space

config = DEFAULT_CONFIG.copy()
config["num_workers"] = 48
config["num_gpus"] = 1
config["model"] = {
    "use_lstm": True,
    "lstm_use_prev_action": True,
    "lstm_use_prev_reward": True,
    "lstm_cell_size": 64,
    "fcnet_hiddens":[],
}
config["framework"] = 'tf'
config["multiagent"] = { "policies": { "a3c_policy": (None, observation_space, action_space, {}), }, "policy_mapping_fn": (lambda agent_id: "a3c_policy")}
config["env_config"] = {}

ray.init()
trainer = A3CTrainer(config=config, env="custom_env")

for i in range(100000):
    result = trainer.train()
    print(f"iteration={i}, episode_reward_mean={result['episode_reward_mean']}")

# save the actor model
checkpoint_path = trainer.save("a3c_"+str(result['episode_reward_mean'])+".checkpoint")
print(f"Model saved at {checkpoint_path}")

-----------------------------------------------------

my_restored_policy = Policy.from_checkpoint("a3c.checkpoint/checkpoint_100000/policies/default_policy")

env=Env("")

initial_state= my_restored_policy.get_initial_state()
prev_action=tuple(np.array([0]) for _ in range(12))
prev_reward=0

obs,__ = np.array(env.reset())

for i in range(200000):

    action,new_state,_ = my_restored_policy.compute_single_action(obs,
                                                                  state=initial_state,
                                                                  prev_action= prev_action,
                                                                  prev_reward=prev_reward,
                                                                  explore=False)

    action = clip_actions(action, env.action_space)

    next_obs,reward,_,__,___=env.step(action)
    obs=next_obs
    prev_reward=reward
    prev_action=action
    initial_state = new_state

### Issue Severity

High: It blocks me from completing my task.

**Comment by ArturNiederfahrenhorst** [comment]
@fpartika What version of Ray are you using? Could you please use interface of `Algorithm` to compute actions here and respond how that goes?

**Comment by ArturNiederfahrenhorst** [comment]
I should add here that A3CTrainer is now called A3C. So I'm guessing you are using an older version of RLlib.
So with that in mind, you'd use A3CTrainer.compute_single_action().

**Comment by ArturNiederfahrenhorst** [comment]
The reason for this is that your observations probably lack preprocessing the way you are doing it.

**Comment by fpartika** [comment]
Thanks for your help @ArturNiederfahrenhorst , I tested using A3C from Algorithm (version 2.3.0), since I want the LSTM to get the time-steps as input I disabled fcnet_hiddens”:[] but unfortunately the result of test is still far from training, please see my config:

config[“num_gpus”] = 1
config[“model”] = {
“use_lstm”: True,
“lstm_use_prev_action”: True,
“lstm_use_prev_reward”: True,
“lstm_cell_size”: 64,
“fcnet_hiddens”:[],

I believe that your hypothesis about lack of pre-processing could be the reason, but my understanding is having “fcnet_hiddens”:[] would disable the pre-processing, could you please confirm?

**Comment by ArturNiederfahrenhorst** [comment]
@fpartika `fcnet_hiddens` has nothing to do with preprocessing. `fcnet_hiddens` describes the dimensions of the fully connected layers in the models that RLlib builds.

**Comment by fpartika** [comment]
You are right @ArturNiederfahrenhorst, my statement before was incomplete, I meant the “fcnet_hiddens”:[] together with "_disable_action_flattening"=True yield the 2D observation going directly to the LSTM, since using fcnet_hiddens would create a fully connected layer before the LSTM by my knowledge, please comment.

Please see my training code below using ray 2.3.0 for completeness of my statement:

import ray
from ray.rllib.algorithms.a3c import A3C, DEFAULT_CONFIG
from ray.tune.registry import register_env
import gymnasium as gym

register_env("custom_env", lambda config: Env(config))

env=Env('')
observation_space=env.observation_space
action_space=env.action_space

config = DEFAULT_CONFIG.copy()
config["num_workers"] = 1
config["num_gpus"] = 1
config["_disable_action_flattening"]=True
config["model"] = {
    "use_lstm": True,
    "lstm_use_prev_action": True,
    "lstm_use_prev_reward": True,,
    "lstm_cell_size": 64,#256,
    "fcnet_hiddens":[],
}
config["framework"] = 'tf'

ray.init()
trainer = A3C(config=config, env="custom_env")


for i in range(100000):
    result = trainer.train()
    print(f"iteration={i}, episode_reward_mean={result['episode_reward_mean']}")

# save the actor model
checkpoint_path = trainer.save("a3c_"+str(result['episode_reward_mean'])+".checkpoint")
print(f"Model saved at {checkpoint_path}")

**Comment by ArturNiederfahrenhorst** [comment]
@fpartika 
Thanks, I don't see anything wrong with the code you provided.

`_disable_action_flattening` does not disable the flattening of the observation.
You'll need to disable preprocessing and code your own model if you want that to happen.
You can't "throw" any 2D observation into the LSTM cell. RLlib does not provide this functionality and neither to popular implementations of LMST cells afaik.

But all of this does not seem to be connected to your initial problem - why does the trained agent underperform during evaluation? Correct?

**Comment by fpartika** [comment]
Hi @ArturNiederfahrenhorst thanks for clarifying it, you are correct since those options doesn't affect preprocessing I believe the under performance is caused by the lack of preprocessing during the test. Sorry about the confusion, it is not a bug.
Let me clarify what I'm referring as 2D,it is (time steps, observations), since the time distribution of the observations matters for my problem. Could you please point me to the right direction in order to get guidance considering that my use-case is not common, what would be the best channel to get further help?
If you allow me a last question, I found the option config["preprocessor_pref"]=None promising to get my (time steps, observations) thru LSTM without preprocessing. I'm thinking about testing it for a couple days, if you say it makes sense.
Many thanks for your guidance!
